pkg/lexer/state.go

```diff
@@ -71,47 +83,58 @@ next:
 
 	switch {
 	case l.ch == '\n':
-		// semicolon insertion
+		// auto semicolon insertion
```

Comment:
Nice!

Reply:
Lol, for some reason my edited comment didn't show up (maybe I forgot to click update). Here I wanted to write, nice idea, but this should be done at the parser level not the lexer level.

Reply:
I believe the parser should just work with the token stream, not manipulate it. That job is for the lexer.

Reply:
If you really want to add semicolons in the lexer then to do it would be to lex the result twice because you need to know everything about the tokens around where you want to insert the semicolon. just having a new line is not enough to know that you need to add a semicolon - you're language is not simple to handle that case in lexer by lexing once - for example, you have the `{` token so consider this

```
something {
...
}
```
Do you see that the semicolon cannot be added there - if you add a redundant one then you need to make sure that you can handle an empty statement - which is not in your grammar. From what I can see you're only looking at the last token but that only works in simple languages so this strategy is not a good one for this in my opinion. Feel free to do what you want though, up to you.

Reply:
Can you give an example where looking at the last token approach can go wrong?

Reply:
what is going to happen in the above case - you will just add a new token like this

```
something {;
...;
};
```
Do we need the first one that is being added in after `{`? I don't think so.
Where it can go wrong is this

```
let a := \ 
    b == c
``` 

what will happen in this case? (this is a made up example, but anything that can involve more than one/two operator in one statement has the potential to go wrong

Reply:
For the first example, It won't be added after the `{`. `pkg/token` has the following function:
```go
func (tok TokenType) InsertSemi() bool {
	if tok.IsLiteral() {
		return true
	}

	switch tok {
	case RPAREN, RBRACK, RBRACE, BREAK, CONTINUE, RETURN:
		return true
	default:
		return false
	}
}
```
We use that to determine whether a semicolon will be inserted after an operator or not. `LBRACE` is not among them.

For the second example too the same rules apply.

Reply:
Oh ok that makes sense then!